redis_feature_store.js

```diff
@@ -0,0 +1,159 @@
+  }
+
+  store.all = function(cb) {
+    client.hgetall(features_key, function(err, obj) {
+      if (err) {
+        config.logger.error("[LaunchDarkly] Error fetching flag from redis", err)
+        cb(null);
+      } else {
+        var results = {}, 
+            flags = obj;
+
+        for (var key in flags) {
+          if (Object.hasOwnProperty.call(flags,key)) {
+            var flag = JSON.parse(flags[key]);
+            if (!flag.deleted) {
+              results[key] = flag;          
+            }
+          }
+        }
+        cb(results);
+      }
+    });
+  }
+
+  store.init = function(flags, cb) {
+    var multi = client.multi();
+    
+    multi.del(features_key);
+    cache.clear();
+
+    for (var key in flags) {
+      if (Object.hasOwnProperty.call(flags,key)) {
+        multi.hset(features_key, key, JSON.stringify(flags[key]));
+      }
+      if (cache_ttl) {
+        cache.put(key, flags[key], cache_ttl);
+      }
+    }
+
+    multi.exec(function(err, replies) {
+      if (err) {
+        config.logger.error("[LaunchDarkly] Error initializing redis feature store", err);
+      } 
+      cb();
+    });
+  }
+
+  store.delete = function(key, version, cb) {
+    var multi;
+    client.watch(features_key);
+    multi = client.multi();
+
+    // We need to run the get() code in a multi txn
+    do_get(multi, key, function(flag) {
+      if (flag) {
+        if (flag.version >= version) {
+          cb();
+          return;          
+        } else {
+          flag.deleted = true;
+          flag.version = version;
+          multi.hset(features_key, key, JSON.stringify(flag));
+          multi.exec(function(err, replies) {
+            if (err) {
+              config.logger.error("[LaunchDarkly] Error deleting feature flag", err);
+            } else {
+              cache.put(key, flag, cache_ttl);
+            }
+            cb();
+          })
+        }
+      } 
+    });
+  }
+
+  store.upsert = function(key, flag, cb) {
+    var multi;
+    client.watch(features_key);
+    multi = client.multi();
+
+    do_get(multi, key, function(original) {
+      if (original) {
+        if (original.version >= version) {
+          cb();
+          return;          
+        } else {
+          multi.hset(features_key, key, JSON.stringify(flag));
+          multi.exec(function(err, replies) {
+            if (err) {
+              config.logger.error("[LaunchDarkly] Error upserting feature flag", err);
+            } else {
+              cache.put(key, flag, cache_ttl);
+            }
+            cb();
+          })
+        }
+      } 
+    })
+  }
+
+  store.initialized = function(cb) {
+    var init = cache.get('$initialized$');
+
+    if (init) {
+      return true;
+    }    
+
+    client.exists('$initialized$', function(err, obj) {
+      if (!err && obj) {
+        cache.set('$initialized$', true);
+      } 
+      cb(!err && obj);
+    });
+  }
+
+  store.close = function() {
+    client.quit();
```

Comment:
This was a fun one. I spent a couple hours chasing down red herrings on the redis side of things, but it turns out that my driver program *was exiting cleanly*, but after a 30 second delay. That made me realize that the problem was the caching library, as I'd set the cache TTL to 30 seconds. So turns out that the caching library doesn't give you an option to dispose of the timer, and any program will wait for the last timer to run and flush cache entries.

I switched to an alternative caching library that has better disposal options.